core/src/Cache/Keys/MsalAccessTokenCacheKey.cs

```diff
@@ -0,0 +1,167 @@
+﻿//------------------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+using System.Text;
+
+namespace Microsoft.Identity.Core.Cache
+{
+    /// <summary>
+    /// An object representing the key of the token cache AT dictionary. The 
+    /// format of the key is not important for this library, as long as it is unique.
+    /// </summary>
+    internal class MsalAccessTokenCacheKey
```

Comment:
This is just a refactoring + move of an existing class. I decided to remove the inheritance hierarchy because the Keys are just data classes: 

- the 3 level hierarchy makes it more difficult to understand what fields go in each key
- we had a "type" enum in the base class. this is a code smell that inheritance is not a good decision here. 

The spec does mention that all keys share "env" and "clientid". We don't use this detail in our code, but we can just as easy extract an interface.